sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1910,486 @@ def transform(self, X):
+
+        if sparse.issparse(X):
+            self._sparse_fit(X, rng)
+        else:
+            self._dense_fit(X, rng)
+
+        return self
+
+    def _transform_col(self, X_col, feature_idx, inverse):
+        """Private function to transform a single feature"""
+
+        output_distribution = getattr(stats, self.output_distribution)
+        # older version of scipy do not handle tuple as fill_value
+        # clipping the value before transform solve the issue
+        if not inverse:
+            lower_bound_x = self.quantiles_[0, feature_idx]
+            upper_bound_x = self.quantiles_[-1, feature_idx]
+            lower_bound_y = 0
+            upper_bound_y = 1
+        else:
+            lower_bound_x = 0
+            upper_bound_x = 1
+            lower_bound_y = self.quantiles_[0, feature_idx]
+            upper_bound_y = self.quantiles_[-1, feature_idx]
+            #  for inverse transform, match a uniform PDF
+            X_col = output_distribution.cdf(X_col)
+        # find index for lower and higher bounds
+        lower_bounds_idx = (X_col - BOUNDS_THRESHOLD <
+                            lower_bound_x)
+        upper_bounds_idx = (X_col + BOUNDS_THRESHOLD >
+                            upper_bound_x)
+
+        references = np.linspace(0, 1, self.n_quantiles, endpoint=True)
+        if not inverse:
+            X_col = np.interp(X_col, self.quantiles_[:, feature_idx],
+                              references)
+        else:
+            X_col = np.interp(X_col, references,
+                              self.quantiles_[:, feature_idx])
+
+        X_col[upper_bounds_idx] = upper_bound_y
+        X_col[lower_bounds_idx] = lower_bound_y
+        # for forward transform, match the output PDF
+        if not inverse:
+            X_col = output_distribution.ppf(X_col)
+            # find the value to clip the data to avoid mapping to
+            # infinity. Clip such that the inverse transform will be
+            # consistent
+            clip_min = output_distribution.ppf(BOUNDS_THRESHOLD -
+                                               np.spacing(1))
+            clip_max = output_distribution.ppf(1 - (BOUNDS_THRESHOLD -
+                                                    np.spacing(1)))
+            X_col = np.clip(X_col, clip_min, clip_max)
+
+        return X_col
+
+    def _dense_transform(self, X, inverse=False):
+        """Forward and inverse transform for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+
+        inverse : bool, optional (default=False)
+            If False, apply forward transform. If True, apply
+            inverse transform.
+
+        Returns
+        -------
+      X : ndarray, shape (n_samples, n_features)
+            Projected data        """
+
+        for feature_idx in range(X.shape[1]):
+            X[:, feature_idx] = self._transform_col(X[:, feature_idx],
```

Comment:
do you mean to be modifying the data in-place? surely we need a `copy` parameter before doing this.